pixelgl/shader.go

```diff
@@ -0,0 +1,162 @@
+package pixelgl
+
+import (
+	"github.com/faiface/glhf"
+	"github.com/faiface/mainthread"
+	"github.com/go-gl/mathgl/mgl32"
+	"github.com/pkg/errors"
+)
+
+// GLShader is a type to assist with managing a canvas's underlying
+// shader configuration. This allows for customization of shaders on
+// a per canvas basis.
+type (
+	GLShader struct {
+		s      *glhf.Shader
+		vf, uf glhf.AttrFormat
+		vs, fs string
+
+		uniforms []gsUniformAttr
+
+		uniformDefaults struct {
+			transform mgl32.Mat3
+			colormask mgl32.Vec4
+			bounds    mgl32.Vec4
+			texbounds mgl32.Vec4
+		}
+	}
+
+	gsUniformAttr struct {
+		Name  string
+		Type  AttrType
+		Value interface{}
+	}
+)
+
+// reinitialize GLShader data and recompile the underlying gl shader object
+func (gs *GLShader) update() {
+	gs.uf = nil
+	for _, u := range gs.uniforms {
+		gs.uf = append(gs.uf, glhf.Attr{
+			Name: u.Name,
+			Type: glhf.AttrType(u.Type),
+		})
+	}
+	var shader *glhf.Shader
+	mainthread.Call(func() {
+		var err error
+		shader, err = glhf.NewShader(
+			gs.vf,
+			gs.uf,
+			gs.vs,
+			gs.fs,
+		)
+		if err != nil {
+			panic(errors.Wrap(err, "failed to create Canvas, there's a bug in the shader"))
+		}
+	})
+
+	gs.s = shader
+}
+
+// gets the uniform index from GLShader
+func (gs *GLShader) getUniform(Name string) int {
+	for i, u := range gs.uniforms {
+		if u.Name == Name {
+			return i
+		}
+	}
+	return -1
+}
+
+// AddUniform appends a custom uniform name and value to the shader
+//
+// To add a time uniform for example:
+//
+// utime := float32(time.Since(starttime)).Seconds())
+// mycanvas.shader.AddUniform("u_time", &utime)
```

Comment:
instead of supporting pointer types, I'd be for adding a `SetUniform` method to change the value of a uniform